### PR TITLE
Fail on unrecognized arguments

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mok@vmware.com>

xref: https://github.com/kubernetes/website/issues/20623
@neolit123 

/kind bug

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks, `NoArgs` should work fine.
/lgtm
/priority backlog

File: cmd/kubeadm/app/cmd/cmd.go [hunk 1]
```diff
@@ -77,6 +77,7 @@ func NewKubeadmCommand(in io.Reader, out, err io.Writer) *cobra.Command {
 			}
 			return nil
 		},
+		Args: cobra.NoArgs,
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/cmd.go:80]: actually i think we shouldn't add this to the parent kubeadm command as it propagates to all sub-commands and we do have args in places.

**Comment by neolit123** [comment]
do we need this for the kubelet too?

File: cmd/kubeadm/app/cmd/cmd.go (hunk 1 above)
Comment (enj) [reply cmd/kubeadm/app/cmd/cmd.go:80]: Ah yeah I missed the `AddCommand`s below.  I was going with a soft rule of "no `AddCommand`s and no `args` processing."

**Comment by enj** [comment]
> do we need this for the kubelet too?

https://github.com/kubernetes/kubernetes/blob/f8297af392a41d5cc9b029766c3ec64cfc908df3/cmd/kubelet/app/server.go#L147-L157

scared me away.

**Comment by jpbetz** [comment]
Strictly speaking, this change is backward incompatible, right? If I'm reading https://github.com/spf13/cobra#positional-and-custom-arguments correctly, if a caller previously had a positional argument on one of these commands that should not have any positional arguments, the positional argument was ignored.

**Comment by neolit123** [comment]
> Strictly speaking, this change is backward incompatible, right? If I'm reading https://github.com/spf13/cobra#positional-and-custom-arguments correctly, if a caller previously had a positional argument on one of these commands that should not have any positional arguments, the positional argument was ignored.

that is true. so if we really want the change long term, we need to convert these to warnings for a period of N releases, mark this PR with `action-required` and announce to k-dev (+other channels) then after N release hard transition.

for the kubeadm sub-commands (e.g. under alpha) we are touching this is not much of an issue, but for the control-plane components it can be.

**Comment by lavalamp** [comment]
I agree with Joe, I don't think you can do this for control plane binaries at least. I know less about kubeadm.

**Comment by enj** [comment]
While I agree this is not exactly backwards compatible, if you are setting flags that are being silently ignored (per the linked issue), you are risking the API server not doing something that you told it to do.

I suspect that a very small subset of clusters will face any issue on an upgrade, though we have no metrics to help us validate that.

Since we have agreed to limit any action-required release notes for v1.19, I am not sure what the best course of action is.

**Comment by lavalamp** [comment]
I agree that it's not good to ignore bogus flags. But I still think a
deprecation process is probably needed. If you can convince @liggitt then
I'm fine with this, otherwise...

**Comment by liggitt** [comment]
The consequences of a flag typo can be *really* severe... for example, running the apiserver with `–authorization-mode=Node,RBAC` could actually run with the default `AllowAll` authorizer.

If someone has an invalid invocation that is dropping args, I think it is better to let them know immediately so they can correct it.

The NoArgs message about an unknown command is pretty confusing... I would probably provide a more helpful message:

```go
func NoArgs(cmd *Command, args []string) error {
	if len(args) > 0 {
		return fmt.Errorf("%q does not take any arguments, got %q", cmd.CommandPath(), args[0])
	}
	return nil
}
```

**Comment by neolit123** [comment]
hello, after seeing [another case](https://github.com/kubernetes/website/issues/21040#issuecomment-631093298) where the k8s website converting `--`  to single unicode dash and `yum` not doing an `exit > 0` for `–disableexcludes` (`–` = `0x8211`) , i was reminded of this PR.

@liggitt as per your comment:
> The consequences of a flag typo can be really severe...

my understanding is that we should proceed with a hard-change + action-required?

@enj please let me know if you are occupied ATM and i can send a new PR with the suggested change by @liggitt

**Comment by liggitt** [comment]
> @liggitt as per your comment:
> 
> > The consequences of a flag typo can be really severe...
> 
> my understanding is that we should proceed with a hard-change + action-required?

that's my recommendation

**Comment by neolit123** [comment]
sent followup here:
https://github.com/kubernetes/kubernetes/pull/91349

**Comment by neolit123** [comment]
this merged https://github.com/kubernetes/kubernetes/pull/91349
/close

**Pull request closed without merging**